main.go

```diff
@@ -0,0 +1,335 @@
+			[]string{"name"},
+			nil,
+		),
+		requestsProcessed: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "", "requests_processed"),
+			"Number of processes served by a process.",
+			[]string{"name", "pid"},
+			nil,
+		),
+		procUptime: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "", "proc_uptime"),
+			"Number of seconds since processor started.",
+			[]string{"name", "pid", "codeRevision"},
+			nil,
+		),
+		procMemory: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "", "proc_memory"),
+			"Memory consumed by a process",
+			[]string{"name", "pid"},
+			nil,
+		),
+	}
+}
+
+// Collect fetches the statistics from the configured passenger frontend, and
+// delivers them as Prometheus metrics. It implements prometheus.Collector.
+func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
+	info, err := e.status()
+	if err != nil {
+		ch <- prometheus.MustNewConstMetric(e.up, prometheus.GaugeValue, 0)
+		log.Errorf("failed to collect status from passenger: %s", err)
+		return
+	}
+	ch <- prometheus.MustNewConstMetric(e.up, prometheus.GaugeValue, 1)
+
+	ch <- prometheus.MustNewConstMetric(e.version, prometheus.GaugeValue, 1, info.PassengerVersion)
+
+	ch <- prometheus.MustNewConstMetric(e.toplevelQueue, prometheus.CounterValue, parseFloat(info.TopLevelRequestsInQueue))
+	ch <- prometheus.MustNewConstMetric(e.maxProcessCount, prometheus.GaugeValue, parseFloat(info.MaxProcessCount))
+	ch <- prometheus.MustNewConstMetric(e.currentProcessCount, prometheus.GaugeValue, parseFloat(info.CurrentProcessCount))
+	ch <- prometheus.MustNewConstMetric(e.appCount, prometheus.GaugeValue, parseFloat(info.AppCount))
+
+	for _, sg := range info.SuperGroups {
+		ch <- prometheus.MustNewConstMetric(e.appQueue, prometheus.GaugeValue, parseFloat(sg.RequestsInQueue), sg.Name)
+		ch <- prometheus.MustNewConstMetric(e.appProcsSpawning, prometheus.GaugeValue, parseFloat(sg.Group.ProcessesSpawning), sg.Name)
+
+		// Update process identifiers map.
+		processIdentifiers = updateProcesses(processIdentifiers, sg.Group.Processes)
+		for _, proc := range sg.Group.Processes {
+			if bucketID, ok := processIdentifiers[proc.PID]; ok {
+				ch <- prometheus.MustNewConstMetric(e.procMemory, prometheus.GaugeValue, parseFloat(proc.RealMemory), sg.Name, strconv.Itoa(bucketID))
+				ch <- prometheus.MustNewConstMetric(e.requestsProcessed, prometheus.CounterValue, parseFloat(proc.RequestsProcessed), sg.Name, strconv.Itoa(bucketID))
+
+				if uptime, err := parsePassengerInterval(proc.Uptime); err == nil {
+					ch <- prometheus.MustNewConstMetric(e.procUptime, prometheus.CounterValue, float64(uptime),
+						sg.Name, strconv.Itoa(bucketID), proc.CodeRevision,
+					)
+				}
+			}
+		}
+	}
+
+}
+
+func (e *Exporter) status() (*Info, error) {
+	var (
+		out bytes.Buffer
+		cmd = exec.Command(e.cmd, e.args...)
+	)
+	cmd.Stdout = &out
+
+	err := cmd.Start()
+	if err != nil {
+		return nil, err
+	}
+
+	errc := make(chan error)
```

Comment:
This channel needs to be buffered, otherwise the goroutine will starve after a timeout.